Masonry repeat(auto-fit) should behave as repeat(auto-fill)

stwrt opened the pull request:
#### f91e41a89ed67eb51626018c4c485931375876f5
<pre>
Masonry repeat(auto-fit) should behave as repeat(auto-fill)
<a href="https://bugs.webkit.org/show_bug.cgi?id=248577">https://bugs.webkit.org/show_bug.cgi?id=248577</a>

Reviewed by Brent Fulgham.

Masonry repeat(auto-fit) should have the properities of repeat(auto-fill) according to the spec.
<a href="https://drafts.csswg.org/css-grid-3/#repeat-auto-fit">https://drafts.csswg.org/css-grid-3/#repeat-auto-fit</a>

Update WPT test case for auto-fit to behave the same as auto-fill.

* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/masonry/tentative/masonry-grid-template-columns-computed-withcontent.html:
* Source/WebCore/rendering/RenderGrid.cpp:
(WebCore::RenderGrid::computeEmptyTracksForAutoRepeat const):
(WebCore::RenderGrid::isMasonryRows const):
(WebCore::RenderGrid::isMasonryColumns const):
(WebCore::RenderGrid::autoRepeatColumnsType const):
(WebCore::RenderGrid::autoRepeatRowsType const):
(WebCore::RenderGrid::placeItemsOnGrid):
(WebCore::RenderGrid::gridAreaBreadthForChildIncludingAlignmentOffsets const):
(WebCore::RenderGrid::columnAxisOffsetForChild const):
(WebCore::RenderGrid::rowAxisOffsetForChild const):
(WebCore::isMasonryRows): Deleted.
(WebCore::isMasonryColumns): Deleted.
* Source/WebCore/rendering/RenderGrid.h:

Canonical link: <a href="https://commits.webkit.org/257300@main">https://commits.webkit.org/257300@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/29d0fc882cc740220565d6cdc56a2a0baf62e46f

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/98469 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/7678 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/31602 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/107898 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/168168 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/102414 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/8194 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/85070 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/91018 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/104556 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/104136 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/6219 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/89777 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/33211 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/88036 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/21136 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/76153 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/1620 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/22665 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/1542 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/45165 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5027 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/6465 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/42086 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/2917 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

stwrt commented on LayoutTests/imported/w3c/web-platform-tests/css/css-grid/masonry/tentative/masonry-grid-template-columns-computed-withcontent.html:
```diff
@@ -77,7 +77,7 @@
 test_computed_value("grid-template-columns", 'repeat(auto-fill, 200px)', '200px');
 test_computed_value("grid-template-columns", 'repeat(auto-fit, [one] 20%)',
-                    '[one] 60px [one] 0px [one] 0px [one] 0px [one] 0px');
+                    '[one] 60px [one] 60px [one] 60px [one] 60px [one] 60px');
```
Want some feedback on this. We are supposed to treat it as an auto-fill, so this seems to make sense to me.

mattwoodrow replied:
I agree that this is a correct change to make, the tracks have a definite size and aren't being collapsed (since it's auto-fill, not auto-fit).

You'll need to upstream this change to WPT once it merges.

Loirooriol commented on Source/WebCore/rendering/RenderGrid.cpp:
```diff
@@ -698,16 +698,45 @@ static GridArea insertIntoGrid(Grid& grid, RenderBox& child, const GridArea& are
+{
+    auto autoRepeatRow = style().gridAutoRepeatRowsType();
+
+    if (style().gridMasonryColumns() && autoRepeatRow == AutoRepeatType::Fit)
```
Should this be `isMasonryColumns()` instead of `style().gridMasonryColumns()`?

stwrt replied:
Nice catch. That sounds correct based on the spec.

The substitution of values on certain behavior has made this quite a lot more complicated.

brentfulgham commented on Source/WebCore/rendering/RenderGrid.cpp:
```diff
@@ -698,16 +698,45 @@ static GridArea insertIntoGrid(Grid& grid, RenderBox& child, const GridArea& are
 }
 
-static inline bool isMasonryRows(const RenderStyle& style)
+bool RenderGrid::isMasonryRows() const
```
Nit: This naming is awkward. Really should be `isMasonryRow` or `areMasonryRows` or something. But not worth changing this patch for.

brentfulgham commented on Source/WebCore/rendering/RenderGrid.cpp:
```diff
@@ -698,16 +698,45 @@ static GridArea insertIntoGrid(Grid& grid, RenderBox& child, const GridArea& are
+// Masonry Spec Section 2
+// "If masonry is specified for both grid-template-columns and grid-template-rows, then the used value for grid-template-columns is none,
+// and thus the inline axis will be the grid axis."
+
```
No whitespace between method comment and method definition, please.

stwrt replied:
Fixed

brentfulgham commented on Source/WebCore/rendering/RenderGrid.cpp:
```diff
@@ -698,16 +698,45 @@ static GridArea insertIntoGrid(Grid& grid, RenderBox& child, const GridArea& are
+// Masonry Spec Section 2.3.1 repeat(auto-fit)
+// "repeat(auto-fit) behaves as repeat(auto-fill) when the other axis is a masonry axis."
+// We need to lie here that we are are really an auto-fill instead of an auto-fit.
+
```
Ditto: We don't use this whitespace between method comment and method.

nt1m replied:
There is a typo here btw "are are"

stwrt replied:
Fixed

brentfulgham reviewed (approved):
Looks good! A few minor nits.

stwrt committed:
Masonry repeat(auto-fit) should behave as repeat(auto-fill)
https://bugs.webkit.org/show_bug.cgi?id=248577

Reviewed by Brent Fulgham.

Masonry repeat(auto-fit) should have the properities of repeat(auto-fill) according to the spec.
https://drafts.csswg.org/css-grid-3/#repeat-auto-fit

Update WPT test case for auto-fit to behave the same as auto-fill.

* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/masonry/tentative/masonry-grid-template-columns-computed-withcontent.html:
* Source/WebCore/rendering/RenderGrid.cpp:
(WebCore::RenderGrid::computeEmptyTracksForAutoRepeat const):
(WebCore::RenderGrid::isMasonryRows const):
(WebCore::RenderGrid::isMasonryColumns const):
(WebCore::RenderGrid::autoRepeatColumnsType const):
(WebCore::RenderGrid::autoRepeatRowsType const):
(WebCore::RenderGrid::placeItemsOnGrid):
(WebCore::RenderGrid::gridAreaBreadthForChildIncludingAlignmentOffsets const):
(WebCore::RenderGrid::columnAxisOffsetForChild const):
(WebCore::RenderGrid::rowAxisOffsetForChild const):
(WebCore::isMasonryRows): Deleted.
(WebCore::isMasonryColumns): Deleted.
* Source/WebCore/rendering/RenderGrid.h:

Canonical link: https://commits.webkit.org/257300@main
